src/sdk/namespace_fs.js

```
@@ -924,22 +924,41 @@ class NamespaceFS {
-                    // find dir object content file path  and return its stat + xattr of its parent directory
-                    const dir_content_path = await this._find_version_path(fs_context, params);
-                    const dir_content_path_stat = await nb_native().fs.stat(fs_context, dir_content_path);
-                    const xattr = stat.xattr;
-                    stat = { ...dir_content_path_stat, xattr };
+            for (;;) {
+                try {
+                    file_path = await this._find_version_path(fs_context, params, true);
+                    await this._check_path_in_bucket_boundaries(fs_context, file_path);
+                    await this._load_bucket(params, fs_context);
+                    stat = await nb_native().fs.stat(fs_context, file_path);
+
+                    isDir = native_fs_utils.isDirectory(stat);
+                    if (isDir) {
+                        if (!stat.xattr?.[XATTR_DIR_CONTENT] || !params.key.endsWith('/')) {
+                            throw error_utils.new_error_code('ENOENT', 'NoSuchKey');
+                        } else if (stat.xattr?.[XATTR_DIR_CONTENT] !== '0') {
+                            // find dir object content file path  and return its stat + xattr of its parent directory
+                            const dir_content_path = await this._find_version_path(fs_context, params);
+                            const dir_content_path_stat = await nb_native().fs.stat(fs_context, dir_content_path);
+                            const xattr = stat.xattr;
+                            stat = { ...dir_content_path_stat, xattr };
+                        }
+                    }
+                    if (params.version_id && !this._is_versioning_disabled() && this._get_version_id_by_xattr(stat) !== params.version_id) {
+                        dbg.warn('NamespaceFS.read_object_md mismatch version_id', params.version_id, this._get_version_id_by_xattr(stat));
```

can you instead just throw an error and add the error code to the should_retry_link_unlink() check?
with this change, we will handle the retries once